In [134]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Data Preparation

### Loading Both Datasets

df1 = fear_greed_index.csv  
df2 = historical_data.csv

In [135]:
df1 = pd.read_csv('fear_greed_index.csv')
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2644 entries, 0 to 2643
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   timestamp       2644 non-null   int64 
 1   value           2644 non-null   int64 
 2   classification  2644 non-null   object
 3   date            2644 non-null   object
dtypes: int64(2), object(2)
memory usage: 82.8+ KB


In [136]:
print(df1.duplicated().sum())

0


No missing values found in df1  
No duplicates found in df1

In [137]:
df2 = pd.read_csv('historical_data.csv')
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 211224 entries, 0 to 211223
Data columns (total 16 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Account           211224 non-null  object 
 1   Coin              211224 non-null  object 
 2   Execution Price   211224 non-null  float64
 3   Size Tokens       211224 non-null  float64
 4   Size USD          211224 non-null  float64
 5   Side              211224 non-null  object 
 6   Timestamp IST     211224 non-null  object 
 7   Start Position    211224 non-null  float64
 8   Direction         211224 non-null  object 
 9   Closed PnL        211224 non-null  float64
 10  Transaction Hash  211224 non-null  object 
 11  Order ID          211224 non-null  int64  
 12  Crossed           211224 non-null  bool   
 13  Fee               211224 non-null  float64
 14  Trade ID          211224 non-null  float64
 15  Timestamp         211224 non-null  float64
dtypes: bool(1), float64(

In [138]:
print(df2.duplicated().sum())

0


No missing values found in df2  
No duplicates found in df2

In [139]:
df2.describe()

Execution Price   Size Tokens      Size USD  Start Position  \
count    211224.000000  2.112240e+05  2.112240e+05    2.112240e+05   
mean      11414.723350  4.623365e+03  5.639451e+03   -2.994625e+04   
std       29447.654868  1.042729e+05  3.657514e+04    6.738074e+05   
min           0.000005  8.740000e-07  0.000000e+00   -1.433463e+07   
25%           4.854700  2.940000e+00  1.937900e+02   -3.762311e+02   
50%          18.280000  3.200000e+01  5.970450e+02    8.472793e+01   
75%         101.580000  1.879025e+02  2.058960e+03    9.337278e+03   
max      109004.000000  1.582244e+07  3.921431e+06    3.050948e+07   

          Closed PnL      Order ID            Fee      Trade ID     Timestamp  
count  211224.000000  2.112240e+05  211224.000000  2.112240e+05  2.112240e+05  
mean       48.749001  6.965388e+10       1.163967  5.628549e+14  1.737744e+12  
std       919.164828  1.835753e+10       6.758854  3.257565e+14  8.689920e+09  
min   -117990.104100  1.732711e+08      -1.175712  0.000000e+00  1.680000e+12  
25%         0.000000  5.983853e+10       0.016121  2.810000e+14  1.740000e+12  
50%         0.000000  7.442939e+10       0.089578  5.620000e+14  1.740000e+12  
75%         5.792797  8.335543e+10       0.393811  8.460000e+14  1.740000e+12  
max    135329.090100  9.014923e+10     837.471593  1.130000e+15  1.750000e+12

### Converting Timestamps and aligning datasets by date

In [140]:
df1['date'] = pd.to_datetime(df1['date']).dt.normalize()
df2['date'] = pd.to_datetime(df2['Timestamp IST'], format="%d-%m-%Y %H:%M").dt.normalize()

In [141]:
df_merged = df2.merge(df1[['date','classification']], on='date', how='left')
df_merged.head()

Account  Coin  Execution Price  \
0  0xae5eacaf9c6b9111fd53034a602c192a04e082ed  @107           7.9769   
1  0xae5eacaf9c6b9111fd53034a602c192a04e082ed  @107           7.9800   
2  0xae5eacaf9c6b9111fd53034a602c192a04e082ed  @107           7.9855   
3  0xae5eacaf9c6b9111fd53034a602c192a04e082ed  @107           7.9874   
4  0xae5eacaf9c6b9111fd53034a602c192a04e082ed  @107           7.9894   

   Size Tokens  Size USD Side     Timestamp IST  Start Position Direction  \
0       986.87   7872.16  BUY  02-12-2024 22:50        0.000000       Buy   
1        16.00    127.68  BUY  02-12-2024 22:50      986.524596       Buy   
2       144.09   1150.63  BUY  02-12-2024 22:50     1002.518996       Buy   
3       142.98   1142.04  BUY  02-12-2024 22:50     1146.558564       Buy   
4         8.73     69.75  BUY  02-12-2024 22:50     1289.488521       Buy   

   Closed PnL                                   Transaction Hash     Order ID  \
0         0.0  0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   
1         0.0  0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   
2         0.0  0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   
3         0.0  0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   
4         0.0  0xec09451986a1874e3a980418412fcd0201f500c95bac...  52017706630   

   Crossed       Fee      Trade ID     Timestamp       date classification  
0     True  0.345404  8.950000e+14  1.730000e+12 2024-12-02  Extreme Greed  
1     True  0.005600  4.430000e+14  1.730000e+12 2024-12-02  Extreme Greed  
2     True  0.050431  6.600000e+14  1.730000e+12 2024-12-02  Extreme Greed  
3     True  0.050043  1.080000e+15  1.730000e+12 2024-12-02  Extreme Greed  
4     True  0.003055  1.050000e+15  1.730000e+12 2024-12-02  Extreme Greed

### Daily PnL and Number of Trades of each Trader

In [143]:
# Daily pnl per trader
daily_pnl = df_merged.groupby(['Account','date','classification'])['Closed PnL'].sum().reset_index(name='total_pnl')

# Number of trades per trader
daily_trades = df_merged.groupby(['Account','date']).size().reset_index(name='num_trades')

daily_trader_data = daily_pnl.merge(daily_trades, on=['date','Account'])
daily_trader_data

Account       date classification  \
0     0x083384f897ee0f19899168e3b1bec365f52a9012 2024-11-11  Extreme Greed   
1     0x083384f897ee0f19899168e3b1bec365f52a9012 2024-11-17  Extreme Greed   
2     0x083384f897ee0f19899168e3b1bec365f52a9012 2024-11-18  Extreme Greed   
3     0x083384f897ee0f19899168e3b1bec365f52a9012 2024-11-22  Extreme Greed   
4     0x083384f897ee0f19899168e3b1bec365f52a9012 2024-11-26  Extreme Greed   
...                                          ...        ...            ...   
2335  0xbee1707d6b44d4d52bfe19e41f8a828645437aab 2025-04-27          Greed   
2336  0xbee1707d6b44d4d52bfe19e41f8a828645437aab 2025-04-28        Neutral   
2337  0xbee1707d6b44d4d52bfe19e41f8a828645437aab 2025-04-29          Greed   
2338  0xbee1707d6b44d4d52bfe19e41f8a828645437aab 2025-04-30          Greed   
2339  0xbee1707d6b44d4d52bfe19e41f8a828645437aab 2025-05-01        Neutral   

         total_pnl  num_trades  
0         0.000000         177  
1         0.000000          68  
2         0.000000          40  
3    -21227.000000          12  
4      1603.100000          27  
...            ...         ...  
2335   1709.194807          82  
2336   4008.588908         430  
2337   8561.771838         902  
2338   2520.773814          75  
2339   1364.022527         131  

[2340 rows x 5 columns]

# Predictive Model #2

### Features and Labels

In [150]:
features = [ 'classification', 'rolling_win_rate_5d', 'rolling_mean_pnl_5d', 'rolling_std_pnl_5d', 'rolling_avg_trades_5d', 'yesterday_pnl' ]
labels = ['target']

# Sorting Values
daily_trader_data = daily_trader_data.sort_values(['Account','date'])

# rolling_win_rate_5d
daily_trader_data['win'] = daily_trader_data['total_pnl'].transform(lambda x: 1 if x>0 else 0)
daily_trader_data['rolling_win_rate_5d'] = daily_trader_data.groupby('Account')['win'].transform(lambda x: x.rolling(5, min_periods=1).mean())
# Note: Here, a no profit no loss day is considered as no win day, unlike elsewhere

# rolling_mean_pnl_5d
daily_trader_data['rolling_mean_pnl_5d'] = daily_trader_data.groupby(['Account'])['total_pnl'].transform(lambda x: x.rolling(5, min_periods=1).mean())

# rolling_std_pnl_5d
daily_trader_data['rolling_std_pnl_5d'] = daily_trader_data.groupby('Account')['total_pnl'].transform(lambda x: x.rolling(5, min_periods=1).std())

# rolling_avg_trades_5d
daily_trader_data['rolling_avg_trades_5d'] = daily_trader_data.groupby('Account')['num_trades'].transform(lambda x: x.rolling(5, min_periods=1).mean())

# yesterday_pnl
daily_trader_data['yesterday_pnl'] = daily_trader_data.groupby('Account')['total_pnl'].shift(1)

# target
daily_trader_data['target'] = daily_trader_data.groupby('Account')['win'].shift(-1)

# dropping missing values
daily_trader_data.dropna(inplace=True)

### Data Partition

In [154]:
cutoff_date = daily_trader_data['date'].quantile(0.8)

train = daily_trader_data[daily_trader_data['date']<=cutoff_date]
test = daily_trader_data[daily_trader_data['date']>cutoff_date]

x_train = train[features]
x_test = test[features]
y_train = train[labels]
y_test = test[labels]

print(cutoff_date)

2025-03-26 00:00:00


### Data Preprocessing

In [152]:
from sklearn.preprocessing import StandardScaler, OrdinalEncoder, LabelEncoder
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer

num_attrbs = [ 'rolling_win_rate_5d', 'rolling_mean_pnl_5d', 'rolling_std_pnl_5d', 'rolling_avg_trades_5d', 'yesterday_pnl' ]
cat_attrbs = ['classification']

X_preprocessor = ColumnTransformer([
    ('num', StandardScaler(), num_attrbs),
    ('cat', OrdinalEncoder(), cat_attrbs)
])
x_train = X_preprocessor.fit_transform(x_train)
x_test = X_preprocessor.fit_transform(x_test)

le = LabelEncoder()
y_train = le.fit_transform(y_train)
y_test = le.fit_transform(y_test)

c:\Users\divya\anaconda3\Lib\site-packages\sklearn\preprocessing\_label.py:114: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
c:\Users\divya\anaconda3\Lib\site-packages\sklearn\preprocessing\_label.py:114: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


### Training and Prediction

In [153]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr = lr.fit(x_train, y_train)
predictions = lr.predict(x_test)

from sklearn.metrics import roc_auc_score
score = roc_auc_score(y_test, predictions)
print(score)

0.5944579372723086


##### The ROC-AUC Score of our model is ~0.59